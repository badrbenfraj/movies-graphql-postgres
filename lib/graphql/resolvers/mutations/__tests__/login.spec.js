const login = require('../login')
const { db } = require('../../../../adapters/db')
const { invalidPassword } = require('../../../../services/login')

jest.mock('../../../../adapters/db', () => ({
  db: {
    one: jest.fn()
  }
}))

jest.mock('../../../../services/login', () => ({
  invalidPassword: jest.fn()
}))

describe('mutations/login', () => {
  beforeEach(() => {
    db.one.mockClear()
  })

  it('should get a specific user', () => {
    login(null, { input: { username: 'batman@waynecorp.org', password: '' } })

    expect(db.one.mock.calls[0][0]).toMatchSnapshot()
  })

  it('should return error if wrong password', async () => {
    db.one.mockImplementation(() => Promise.resolve({ password: 'hej', username: 'batman'}))
    invalidPassword.mockReturnValue(true)

    try {
      await login(null, {
        input: { username: 'batman@waynecorp.org', password: 'robin4ever' }
      })
    } catch (e) {
      expect(e.message).toMatch(/Wrong username or password/)
    }
  })

  it('should return database data', async () => {
    invalidPassword.mockReturnValue(false)
    db.one.mockReturnValue({ name: 'Bruce Wayne', id: 3})

    const response = await login(null, {
      input: { username: 'batman@waynecorp.org', password: '' }
    })

    expect(response).toMatchSnapshot()
  })

  it('should handle errors', async () => {
    db.one.mockImplementation(() => Promise.reject('nope'))

    try {
      await login(null, {
        input: { username: 'batman@waynecorp.org', password: '' }
      })
    } catch (e) {
      expect(e.message).toMatch(/nope/)
    }
  })
})
