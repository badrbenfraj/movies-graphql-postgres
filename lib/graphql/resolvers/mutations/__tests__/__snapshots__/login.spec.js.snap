// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutations/login should get a specific user 1`] = `"SELECT * FROM users WHERE email = 'batman@waynecorp.org'"`;

exports[`mutations/login should issue a token 1`] = `
Object {
  "email": "batman@waynecorp.org",
  "id": 3,
  "name": "Bruce Wayne",
}
`;

exports[`mutations/login should should lowercase username 1`] = `"SELECT * FROM users WHERE email = 'matman@waynecorp.org'"`;

exports[`mutations/login should should trim username 1`] = `"SELECT * FROM users WHERE email = 'batman@waynecorp.org'"`;
